src/Esprima/ParserExtensions.cs

```diff
@@ -15,617 +16,631 @@ static ParserExtensions()
+                && source[4] == 't' && source[5] == 'r' && source[6] == 'i' && source[7] == 'n' && source[8] == 'g')
             {
                 return "substring";
             }
         }
         else if (span == 10)
         {
-            if (source[from] == 'i' && source[from + 1] == 's' && source[from + 2] == 'F' && source[from + 3] == 'u'
-                && source[from + 4] == 'n' && source[from + 5] == 'c' && source[from + 6] == 't' && source[from + 7] == 'i'
-                && source[from + 8] == 'o' && source[from + 9] == 'n')
+            if (source[0] == 'i' && source[1] == 's' && source[2] == 'F' && source[3] == 'u'
+                && source[4] == 'n' && source[5] == 'c' && source[6] == 't' && source[7] == 'i'
+                && source[8] == 'o' && source[9] == 'n')
             {
                 return "isFunction";
             }
         }
         else if (span == 11)
         {
-            if (source[from] == 't' && source[from + 1] == 'o' && source[from + 2] == 'L' && source[from + 3] == 'o'
-                && source[from + 4] == 'w' && source[from + 5] == 'e' && source[from + 6] == 'r' && source[from + 7] == 'C'
-                && source[from + 8] == 'a' && source[from + 9] == 's' && source[from + 10] == 'e')
+            if (source[0] == 't' && source[1] == 'o' && source[2] == 'L' && source[3] == 'o'
+                && source[4] == 'w' && source[5] == 'e' && source[6] == 'r' && source[7] == 'C'
+                && source[8] == 'a' && source[9] == 's' && source[10] == 'e')
             {
                 return "toLowerCase";
             }
         }
         else if (span == 14)
         {
-            if (source[from] == 'h' && source[from + 1] == 'a' && source[from + 2] == 's' && source[from + 3] == 'O'
-                && source[from + 4] == 'w' && source[from + 5] == 'n' && source[from + 6] == 'P' && source[from + 7] == 'r'
-                && source[from + 8] == 'o' && source[from + 9] == 'p' && source[from + 10] == 'e' && source[from + 11] == 'r'
-                && source[from + 12] == 't' && source[from + 13] == 'y')
+            if (source[0] == 'h' && source[1] == 'a' && source[2] == 's' && source[3] == 'O'
+                && source[4] == 'w' && source[5] == 'n' && source[6] == 'P' && source[7] == 'r'
+                && source[8] == 'o' && source[9] == 'p' && source[10] == 'e' && source[11] == 'r'
+                && source[12] == 't' && source[13] == 'y')
             {
                 return "hasOwnProperty";
             }
         }
 
-        var substring = source.Substring(from, span);
+        return null;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static string Slice(this string source, int start, int end, ref StringPool stringPool)
+    {
+        var sourceSpan = source.AsSpan(start, end - start);
+        return TryGetInternedString(sourceSpan) ?? stringPool.GetOrCreate(sourceSpan);
```

Comment:
Couldn't the keywords in TryGetInternedString be part of the stringPool too? That might make things simpler, and probably as fast.

Reply:
Good point. In the case of string pooling it doesn't make sense indeed to hard-code a bunch of keywords. What seems to make sense is to use interned string instances of keywords that are hard-coded anyway (i.e. the keywords defined in the scanner). It would be redundant to duplicate those by placing them into the string pool.